fedbiomed/node/requests/_pending_requests.py

```diff
@@ -0,0 +1,196 @@
+
+        # check if added reply completes some listeners
+        # (note: there should never be more than 1 event `set()` under current assumptions)
+        completed_listeners = self._all_replies()
+        with self._pending_listeners_lock:
+            for completed_listener in completed_listeners:
+                # check: listener may have been removed since tested `_all_replies` as we don't kept the lock
+                if completed_listener in self._pending_listeners:
+                    # wake up waiting listener
+                    self._pending_listeners[completed_listener]['event'].set()
+
+
+    def add_listener(self, request_ids: list[str]) -> int:
+        """Add a listener waiting for replies to multiple requests to nodes.
+
+        Args:
+            request_ids: list of the unique IDs for all requests to wait
+
+        Returns:
+            An `int`, unique listener ID for this listener
+        """
+        with self._pending_listeners_lock:
+            self._id_counter += 1
+            self._pending_listeners[self._id_counter] = {
+                'start_time': time.time(),
+                'event': threading.Event(),
+                'requests': request_ids,
+            }
+
+        return self._id_counter
+
+
+    def wait(self, listener_id: int, timeout: float) -> Tuple[bool, List[InnerMessage]]:
+        """Wait for a registered listener to complete.
+
+        Blocks until all replies for this listener are received or until timeout is reached.
+
+        Args:
+            listener_id: unique ID of this listener
+            timeout: maximum time to wait for replies in seconds
+
+        Returns:
+            A tuple consisting of
+                - a bool set to `True` if all replies for this listener were received, `False` if not
+                - a list of the messages for the received replies
+
+        Raises:
+               FedbiomedNodeToNodeError: timeout has incorrect type or value
+        """
+
+        # Check value for timeout, as bad value may cause hard to detect problems
+        if (not isinstance(timeout, float) and not (isinstance(timeout, int))) \
+                or timeout < 0 or timeout > MAX_PENDING_REPLY_TIMEOUT:
+            raise FedbiomedNodeToNodeError(f"{ErrorNumbers.FB324}: Cannot wait {timeout} seconds. "
+                                           f"Should be float between 0 and {MAX_PENDING_REPLY_TIMEOUT}")
+
+        time_initial = time.time()
+
+        with self._pending_listeners_lock:
+            # Check a listener exists for this id
+            if listener_id not in self._pending_listeners:
+                # don't raise exception, non fatal error
+                logger.error(f"{ErrorNumbers.FB324}: Node tries to wait for replies "
+                             f"of a non existing request-reply: {listener_id}")
+                return False, []
+
+            # Retrieve event object for this listener
+            event = self._pending_listeners[listener_id]['event']
```

Comment:
Correct me if I am wrong but retrieving values from dict is by default thread safe,  and assuming that wait called once for a specific listener id, lock may not be necessary here.